src/app-layer-expectation.c

```diff
@@ -199,9 +204,18 @@ int AppLayerExpectationCreate(Flow *f, int direction, Port src, Port dst,
 
     iexp = IPPairGetStorageById(ipp, g_expectation_id);
     exp->next = iexp;
-    IPPairSetStorageById(ipp, g_expectation_id, exp);
+    if (iexp) {
+        exp->level = iexp->level + 1;
```

Comment:
I don't understand what is being limited here, can you explain? Does this mean we can have just 10 active ftp-data flows per ip-pair? Or is this about expectations that aren't 'consumed' yet by a flow? What happens if 10 expectations are created but only 1-9 are used. Will the 10th block any further expectations from being created? Would like to see these things explained in the commit msg and maybe also at the top of this file.

Reply:
So we don't add the expectation if there is already 10 non converted expectation. But preparing this answer I did realize that this means, an attacker can saturate the expectation list then pass a command. This needs to be implemented with a FIFO.